metadata-ingestion/src/datahub/ingestion/source/glue.py

```diff
@@ -131,44 +151,270 @@ def __init__(self, config: GlueSourceConfig, ctx: PipelineContext):
+            if node_type in ["DataSource", "DataSink"]:
+
+                node_args = {x["Name"]: json.loads(x["Value"]) for x in node["Args"]}
+
+                # if data object is Glue table
+                if "database" in node_args and "table_name" in node_args:
+
+                    full_table_name = (
+                        f"{node_args['database']}.{node_args['table_name']}"
+                    )
+
+                    # we know that the table will already be covered when ingesting Glue tables
+                    node_urn = f"urn:li:dataset:(urn:li:dataPlatform:glue,{full_table_name},{self.env})"
+
+                # if data object is S3 bucket
+                elif node_args.get("connection_type") == "s3":
+
+                    # remove S3 prefix
+                    s3_name = node_args["connection_options"]["path"][5:]
+
+                    if s3_name.endswith("/"):
+                        s3_name = s3_name[:-1]
+
+                    node_urn = (
+                        f"urn:li:dataset:(urn:li:dataPlatform:s3,{s3_name},{self.env})"
+                    )
+
+                    dataset_snapshot = DatasetSnapshot(
+                        urn=node_urn,
+                        aspects=[],
+                    )
+
+                    dataset_snapshot.aspects.append(Status(removed=False))
+                    dataset_snapshot.aspects.append(
+                        OwnershipClass(
+                            owners=[],
+                            lastModified=AuditStampClass(
+                                time=int(time.time() * 1000),
+                                actor="urn:li:corpuser:datahub",
+                            ),
+                        )
+                    )
+                    dataset_snapshot.aspects.append(
+                        DatasetPropertiesClass(
+                            customProperties={k: str(v) for k, v in node_args.items()},
+                            tags=[],
+                        )
+                    )
+
+                    new_dataset_ids.append(node["Id"])
+                    new_dataset_mces.append(
+                        MetadataChangeEvent(proposedSnapshot=dataset_snapshot)
+                    )
+
+                else:
+
+                    raise ValueError(f"Unrecognized Glue data object type: {node_args}")
+
+            # otherwise, a node represents a transformation
+            else:
+                node_urn = mce_builder.make_data_job_urn_with_flow(
+                    flow_urn, job_id=f'{node["NodeType"]}-{node["Id"]}'
+                )
+
+            return {
+                **node,
+                "urn": node_urn,
+                # to be filled in after traversing edges
+                "inputDatajobs": [],
+                "inputDatasets": [],
+                "outputDatasets": [],
+            }
+
+        nodes = {node["Id"]: process_node(node) for node in dataflow_graph["DagNodes"]}
```

Comment:
Maybe its just me, but the nested function makes this slightly difficult to follow

Reply:
Fixed!